File: telemetry/frame.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Tuple

import msgpack


@dataclass(frozen=True)
class TelemetryFrame:
    """
    High-performance state snapshot for UI and network sync.
    Optimized for msgpack serialization.
    """
    v: int = 2  # Protocol version
    t: int = 0  # Tick
    s: Tuple[int, int] = (0, 0)  # Score (Blue, Red)
    b: Tuple[float, float] = (0.0, 0.0)  # Ball Pos X, Y
    bv: Tuple[float, float] = (0.0, 0.0)  # Ball Vel X, Y
    p: List[Dict[str, Any]] = field(default_factory=list)  # Players: [{id, team, pos:[x,y], vel:[v,v], st, tags[]}]
    e: List[Dict[str, Any]] = field(default_factory=list)  # Events
    o: Dict[str, Any] = field(default_factory=dict) # Overlays (Explainability)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "v": self.v,
            "t": self.t,
            "s": list(self.s),
            "b": list(self.b),
            "bv": list(self.bv),
            "p": self.p,
            "e": self.e,
            "o": self.o
        }

    def pack(self) -> bytes:
        return msgpack.packb(self.to_dict(), use_bin_type=True)

    @classmethod
    def unpack(cls, data: bytes) -> TelemetryFrame:
        d = msgpack.unpackb(data, raw=False)
        return cls(
            v=d["v"],
            t=d["t"],
            s=tuple(d["s"]),
            b=tuple(d["b"]),
            bv=tuple(d["bv"]),
            p=d["p"],
            e=d["e"],
            o=d.get("o", {})
        )

File: ui/lib/frame.ts
export type PlayerFrame = {
  id: string;
  team: string;
  pos: [number, number];
  vel: [number, number];
};

export type Frame = {
  tick: number;
  score: [number, number];
  ball: { pos: [number, number]; vel?: [number, number] };
  players: PlayerFrame[];
  overlays?: { attn?: number[][] };
};

export function normalizeFrame(raw: Record<string, unknown>): Frame {
  const scoreRaw = Array.isArray(raw.s) ? raw.s : [0, 0];
  const ballPos = Array.isArray(raw.b) ? raw.b : [300, 200];
  const playersRaw = Array.isArray(raw.p) ? raw.p : [];

  const players = playersRaw.map((p) => {
    const player = p as Record<string, unknown>;
    return {
      id: String(player.id ?? "unknown"),
      team: String(player.team ?? "BLUE"),
      pos: (Array.isArray(player.pos) ? player.pos : [0, 0]) as [number, number],
      vel: (Array.isArray(player.vel) ? player.vel : [0, 0]) as [number, number],
    };
  });

  return {
    tick: Number(raw.t ?? 0),
    score: [Number(scoreRaw[0] ?? 0), Number(scoreRaw[1] ?? 0)],
    ball: { pos: [Number(ballPos[0] ?? 300), Number(ballPos[1] ?? 200)] },
    players,
    overlays: (raw.o as { attn?: number[][] } | undefined) ?? {},
  };
}
